### PR TITLE
Run apt-get update on startup in background

### LINKED ISSUE #810: Output in non-verbose unattended installation
Check where these come from...

```
2020-03-25_21:09:05_UTC [openHABian] Executing 'openhabian-setup.sh unattended'...
2020-03-25_21:09:05_UTC [openHABian] Checking for root privileges... OK
Hit:1 http://raspbian.raspberrypi.org/raspbian stretch InRelease
Hit:2 http://archive.raspberrypi.org/debian stretch InRelease
Reading package lists...
2020-03-25_21:09:27_UTC [openHABian] Loading configuration file '/etc/openhabian.conf'... OK

```

### PR DESCRIPTION
**Pull request by mstormi**
... and wait for it to finish after menu selection.

Fixes #810 

Signed-off-by: Markus Storm <markus.storm@gmx.net>

### CODE REVIEW

**Comment by EliasGabrielsson** [comment]
Like the Idea of this patch Markus.
Will hopefully install a new system this weekend, will try to test it out then.

**Comment by mstormi** [comment]
> Will hopefully install a new system this weekend, will try to test it out then.

Consider patching #681 - #684 in for that test only, too.

File: functions/influxdb+grafana.bash [hunk 1]
```diff
@@ -218,7 +218,6 @@ influxdb_install() {
   influxdb_address="http://localhost:8086"
   influxdb_admin_username="admin"
   if [ ! -f /etc/influxdb/influxdb.conf ]; then
-    cond_redirect apt-get -y install apt-transport-https
```
Comment (ThomDietrich) [inline functions/influxdb+grafana.bash]: Please move to independent PR, with rationale

File: openhabian-setup.sh [hunk 2]
```diff
@@ -87,7 +87,8 @@ if [[ -n "$UNATTENDED" ]]; then
   clean_config_userpw
   frontail_setup
 else
-  apt-get update
+  apt-get -q update >/dev/null 2>&1 &
+  PID_APT=$!
```
Comment (ThomDietrich) [inline openhabian-setup.sh]: Both in one line please

File: openhabian-setup.sh [hunk 3]
```diff
@@ -59,7 +59,7 @@ OLDWD=$(pwd) && cd /opt
 if [[ -n "$UNATTENDED" ]]; then
   # apt/dpkg commands will not try interactive dialogs
   export DEBIAN_FRONTEND=noninteractive
-  apt-get update
+  cond_redirect apt-get update
```
Comment (ThomDietrich) [inline openhabian-setup.sh]: Would also put in it's own PR. Unrelated

File: functions/openhab.bash [hunk 4]
```diff
@@ -21,8 +21,6 @@ Check the \"openHAB Release Notes\" and the official announcements to learn abou
     TESTING=1
   fi
 
-  cond_redirect apt-get -y install apt-transport-https
```
Comment (ThomDietrich) [inline functions/openhab.bash]: Other PR

File: functions/menu.bash [hunk 5]
```diff
@@ -47,10 +47,10 @@ show_main_menu() {
     openhabian_update
 
   elif [[ "$choice" == "02"* ]]; then
-    system_upgrade
+    wait ${PID_APT} 2>/dev/null; system_upgrade
```
Comment (ThomDietrich) [inline functions/menu.bash]: Can we move this into a function with a speaking name, please? It's impossible to maintain this in the long run.

**Comment by ThomDietrich** [comment]
Like this PR overall

File: functions/openhab.bash (hunk 4 above)
Comment (mstormi) [reply functions/openhab.bash]: ok

File: functions/menu.bash (hunk 5 above)
Comment (mstormi) [reply functions/menu.bash]: ok now?

File: openhabian-setup.sh [hunk 6]
```diff
@@ -59,7 +59,7 @@ OLDWD=$(pwd) && cd /opt
 if [[ -n "$UNATTENDED" ]]; then
   # apt/dpkg commands will not try interactive dialogs
   export DEBIAN_FRONTEND=noninteractive
-  apt-get update
+  apt_update
```
Comment (ThomDietrich) [inline openhabian-setup.sh]: Do we need to wait for it to finish here?

Comment (mstormi) [reply openhabian-setup.sh]: not here but before calling apt-get next time which is in timezone routine

File: functions/openhabian.bash [hunk 7]
```diff
@@ -9,6 +9,16 @@ get_git_revision() {
...
 }
 
+apt_update() {
+  echo -n "$(timestamp) [openHABian] Updating Linux package information ... "
+  apt-get -q update >/dev/null 2>&1 &
+  PID_APT=$!
+}
+
+wait_for_apt_to_finish_update() {
+    wait ${PID_APT} 2>/dev/null
```
Comment (holgerfriedrich) [inline functions/openhabian.bash:19]: Pls make sure that PID_APT is set, otherwise call apt_update. Then we can still source the function files directly from shell without changing the experience (wait w/o a pid will just wait for child tasks, if there are none, it will continue).

Comment (holgerfriedrich) [reply functions/openhabian.bash:19]: did you think of using `wait -f``?

Comment (holgerfriedrich) [reply functions/openhabian.bash:19]: Might be a bit strange for users: a) it might wait for long without showing a reason b) on connection errors we are hiding the reason....

Comment (mstormi) [reply functions/openhabian.bash]: ok now check for PID_APT and used wait -f
On your last point I would leave it like that. IIRC it does not work without that redirection so that's the lesser pain point (and only when Inet is down).

File: functions/openhabian.bash [hunk 8]
```diff
@@ -9,6 +9,20 @@ get_git_revision() {
...
+  echo -n "$(timestamp) [openHABian] Updating Linux package information ... "
+  apt-get -q update >/dev/null 2>&1 &
+  PID_APT=$!
+}
+
+wait_for_apt_to_finish_update() {
+  if [ -v PID_APT ]; then
+    wait -f ${PID_APT} 2>/dev/null
+  else
+    apt_update
```
Comment (holgerfriedrich) [inline functions/openhabian.bash:22]: ...and wait for completion
so the better approach would be
`if not ... then apt_update; fi
wait -f ...`

Comment (mstormi) [reply functions/openhabian.bash:20]: right
